codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -254,6 +259,82 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
         }
     }
 
+    @Override
+    protected ChannelPromise encode(ChannelHandlerContext ctx, ByteBuf msg, ByteBuf out, ChannelPromise promise)
+            throws Exception {
+        try {
+            encode(ctx, msg, out);
+        } catch (Exception e) {
+            if (bufferPromise != null) {
+                bufferPromise.tryFailure(e);
+                bufferPromise = null;
+            }
+            throw e;
+        }
+
+        if (bufferPromise == null) {
+            if (finished) {
+                return promise;
+            }
+            if (buffer.isReadable()) {
+                if (out.isReadable()) {
+                    // The second half of msg got buffered in #buffer and the first half went into out so both have to
+                    // succeed for msg's promise to succeed. We create a new promise for the part of the message in
+                    // out because we started without a pending #bufferPromise.
+                    return combinedPromise(ctx, promise, ctx.newPromise());
+                }
+                bufferPromise = promise;
+                // out is empty and we just return a new promise as out does not contain any part of msg
+                return ctx.newPromise();
+            }
+            return promise;
+        }
+        if (buffer.isReadable()) {
+            if (out.isReadable()) {
+                // the second half of msg got buffered in #buffer and the first half went into out so both have to
+                // succeed for msg's promise to succeed so we associate #bufferPromise with out
+                return combinedPromise(ctx, promise, bufferPromise);
+            }
+            bufferPromise.addListener(new ChannelPromiseNotifier(promise));
+            // out is empty and we just return a new promise as out does not contain any part of msg
+            return ctx.newPromise();
+        }
+        // msg exactly filled up a frame so both the previously buffered data as well as msg are now in out so we
+        // notify both simultaneously through #bufferPromise
+        bufferPromise.addListener(new ChannelPromiseNotifier(promise));
+        final ChannelPromise prev = bufferPromise;
+        bufferPromise = null;
+        return prev;
+    }
+
+    /**
+     * This method handles the promise ({@code msgPromise}) to notify for a message that is split between the current
+     * output buffer and {@link #buffer}. In this case {@code msgPromise} is notified with success once the current
```

Comment:
Can you explain more as to why you took this approach of notifying promises as opposed to accumulating all pending promises in `bufferPromise` and complete/fail them together? Why do we have to selectively handle failure of `out` promise?

Reply:
This should be covered by https://github.com/netty/netty/pull/11419#discussion_r663922122 as well I hope.